### ISSUE TITLE
Plugin upgrade version order not guaranteed

### ISSUE THREAD

**Issue by amartin-parkeon** [pf4j/pf4j#248]
On my system (Ubuntu 16.04, SSD), the org.pf4j.LoadPluginsTest#upgrade test fails (see attached logs : 
[LoadPluginsTest.log](https://github.com/pf4j/pf4j/files/2572593/LoadPluginsTest.log) ).

It seems the library is relying on the method j[ava.io.File.listFiles(FilenameFilter filter)](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#listFiles(java.io.FileFilter)) in org.pf4j.BasePluginRepository#getPluginPaths for listing the plugin folders matching a specific plugin. Yet, "[...]The behavior of this method is the same as that of the [listFiles()](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#listFiles()) method [...]"  which gives "[...] no guarantee that the name strings in the resulting array will appear in any specific order; they are not, in particular, guaranteed to appear in alphabetical order.[...]" resulting in the test failing on my system.

Did I miss something or there should be a sort made on the results before passing them to the org.pf4j.AbstractPluginManager#loadPluginFromPath method at org/pf4j/AbstractPluginManager.java:222 ?

**Comment by decebals** [comment]
Can you create a quickstart (tiny) project that replicates the problem and make it public on Github?
I also use Ubuntu so I can quick test the problem.
Thanks!

**Comment by amartin-parkeon** [comment]
Attached is a patch [issue_#248.txt](https://github.com/pf4j/pf4j/files/2576581/issue_.248.txt) that add a test showing the problem on my system with a proposed solution (commented out).

I had to "cheat" a little bit in the test by purposely return result of org.pf4j.BasePluginRepository#getPluginPaths unordered to show the problem. In my case, this is the result of the system call  java.io.File.listFiles(FilenameFilter filter) which is not ordered alphabetically so it's not easy to simulate (I tried with PowerMock but had some issues and no time to investigates).
